luarules/gadgets/unit_collision_damage_behavior.lua

```diff
@@ -161,23 +186,34 @@ function gadget:UnitUnloaded(unitID, unitDefID, unitTeam,  transportID, transpor
 	fallingUnits[unitID] = true --units falling from transports should take collision damagee from any trajectory, including when bouncing off of other objects.
 end
 
+function gadget:UnitEnteredAir(unitID, unitDefID, unitTeam)
+	if not transportedUnits[unitID] and not unitDefData[unitDefID].canFly then
+		launchedUnits[unitID] = true
+	end
+end
+
 function gadget:UnitLeftAir(unitID, unitDefID, unitTeam)
 	fallingUnits[unitID] = nil
+	launchedUnits[unitID] = nil
 end
 
 function gadget:UnitDestroyed(unitID, unitDefID, unitTeam, attackerID, attackerDefID, attackerTeam)
 	transportedUnits[unitID] = nil
 	unitInertiaCheckFlags[unitID] = nil
+	launchedUnits[unitID] = nil
+	fallingUnits[unitID] = nil
 end
 
 function gadget:GameFrame(frame)
-	for unitID, expirationFrame in pairs(unitInertiaCheckFlags) do
+	for unitID, data in pairs(unitInertiaCheckFlags) do
 		if not transportedUnits[unitID] and not spGetUnitIsDead(unitID) then
 			local velX, velY, velZ, velocityLength = spGetUnitVelocity(unitID)
-			
-			if velocityLength > velocityCap then -- Sheer off extreme velocities within acceptable range
+			if not data.velocityReduced and velocityLength > data.velocityCap then -- Sheer off extreme velocities within acceptable range
 				local verticalVelocityCapThreshold = 0.07 --value derived from empirical testing to prevent fall damage and goofy trajectories from impulse
-				local horizontalVelocity = math.sqrt(velX^2 + velZ^2)
+				local horizontalVelocity =  0
+				if velX ~= 0 or velY ~= 0 then
+					horizontalVelocity = math.sqrt(velX^2 + velZ^2)
+				end
```

Comment:
What is the logic here? Horizontal velocity is by definition `math.sqrt(velX^2 + velZ^2)`, why is anything else being done?

Reply:
performing square root on 0 values causes the game to crash with no reported errors. Same kind of weirdness as dividing by 0.

Reply:
That makes absolutely no sense, square root of 0 is 0. Square root of a negative number is what would cause errors, but the calculation here is adding two squared numbers, and a squared number is guaranteed to be positive.

Reply:
Huh. Ain't sure what that's about.
When I did these checks for 0, it stopped the game crashes. That's what I care about.

Reply:
im running a few tests without

Reply:
Maybe when damgam made that change that fixed that crash it also helped with this random crash problem. Since removing the 0 check I made it through a whole barb vs barb game with cranked up impulse values without any crashes which wasn't happening before. Cool.